### DISCUSSION TITLE
posts/linux-kernel-pwn-01/

### DISCUSSION THREAD

**Discussion by giscus** [brant-ruan/fernweh-comments#18] [Announcements]
# posts/linux-kernel-pwn-01/

Linux Kernel PWN | 01 From Zero to One

https://blog.wohin.me/posts/linux-kernel-pwn-01/

**Comment by Icegrave0391** [comment]
Towards your questions:

1. 
`commit_creds(prepare_kernel_cred(0))` is a well-known payload (gadget chain) to for privilege escalation.
Refer to paper [Shard](https://www.usenix.org/system/files/sec21summer_abubakar.pdf)'s Table 6 (P1).
A bunch of real-world PoCs leverage that payload to perform in-kernel ROP to elevate attacker privileges.

2. 
Before Linux 5.3, a common gadget (`mov cr4, xxx`) can be found in the function `native_write_cr4` to bypass SMAP/SMEP. However, after Linux 5.3, that function is patched (refer to this [article](https://0x434b.dev/dabbling-with-linux-kernel-exploitation-ctf-challenges-to-learn-the-ropes/#smepsmap)).
Regardless, is not necessary for attackers to disable SMAP/SMEP today. They can perform pure in-kernel ROP (e.g., through elastic objects like msg_msg; see [2021-22555-writeup](https://github.com/google/security-research/blob/master/pocs/linux/cve-2021-22555/writeup.md#bypassing-smap)), DirtyCred (see [2021-4154-writeup](https://bsauce.github.io/2022/10/17/CVE-2021-4154/)), or even overwrite `modprobe_path` (see [2021-43267-writeup](https://bsauce.github.io/2021/12/06/CVE-2021-43267/)).

3.
(Correct me if I'm wrong) I think stack pivot is basically a combination of stack overwriting + further ROP. See [2022-1015-writeup](https://bsauce.github.io/2022/07/16/CVE-2022-1015/).

**Reply by brant-ruan** [reply]
This is an awesome replying! For the third one, the idea of stack pivoting is used in both kernel and userland exploits.
